Alerting: Add store level pagination of rules

rwwiv opened the pull request:
**What is this feature?**

This PR adds a new `ListByRuleGroup` method to the rule store that allows us to paginate by group while serializing results from the database, and pipes this through to the Prometheus api that previously handled pagination.

**Why do we need this feature?**

This provides significant speedups vs listing and serializing all rules, especially with larger datasets. In local testing on the store method this results in a fairly linearly scaling speed increase (100x at 1% of total rules, 10x at 10%) as shown by the benchmark results below. 

Benchmark (10000 rules, 100 rules per group)

<img width="1000" alt="Benchmark showing results" src="https://github.com/user-attachments/assets/252630c0-6a3a-4863-97a0-9ddc4bed17de" />

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/contribute/release-notes/#how-to-determine-if-content-belongs-in-whats-new), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/contribute/release-notes/) doc.

rwwiv committed:
Add store level pagination of rules

rwwiv committed:
Test pagination works as expected

moustafab reviewed (dismissed):
Looks great! Awesome work!

moustafab reviewed (commented):
Store side looks good, how much is needed to replace the calls on the prometheus api?

rwwiv committed:
Small refactoring

rwwiv committed:
Pipe pagination changes up to API

This also necessitates writing pagination functionality for the fake
rule store. Technically some of the previous tests are pointless now
that they're testing the fake store, but keeping them doesn't hurt
anything.

moustafab commented on pkg/services/ngalert/store/alert_rule.go:
```diff
@@ -582,6 +582,186 @@ func (st DBstore) CountInFolders(ctx context.Context, orgID int64, folderUIDs []
+			}
+			if key != cursor {
+				// Check if we've reached the group limit
+				if query.GroupLimit > 0 && groupsFetched == int(query.GroupLimit) {
```
```suggestion
				if query.GroupLimit > 0 && groupsFetched == query.GroupLimit {
```

moustafab commented on pkg/services/ngalert/tests/fakes/rules.go:
```diff
@@ -185,6 +186,91 @@ func (f *RuleStore) GetAlertRulesGroupByRuleUID(_ context.Context, q *models.Get
+	}
+
+	outputRules := make([]*models.AlertRule, 0, len(ruleList))
+	groupsFetched := 0
```
```suggestion
	var groupsFetched int64
```

moustafab commented on pkg/services/ngalert/tests/fakes/rules.go:
```diff
@@ -185,6 +186,91 @@ func (f *RuleStore) GetAlertRulesGroupByRuleUID(_ context.Context, q *models.Get
+			RuleGroup:    r.RuleGroup,
+		}
+		if key != cursor {
+			if q.GroupLimit > 0 && groupsFetched == int(q.GroupLimit) {
```
```suggestion
			if q.GroupLimit > 0 && groupsFetched == q.GroupLimit {
```

moustafab reviewed (commented):
Small changes on avoiding the int conversion (doesn't matter too much though)

moustafab committed:
clean up int conversion

moustafab reviewed (approved):
Looks good to me, but given that I've touched it I added @fayzal-g to have another set of eyes

moustafab committed:
clean up lint

fayzal-g reviewed (approved):
Nice! Lgtm

rwwiv committed:
Remove vestigial comment

moustafab committed:
remove folder creation from store integration test

moustafab committed:
remove unneeded code from test fixture
